File: dynamic-property-zk/src/test/java/ru/fix/dynamic/property/zk/ZkDynamicPropertySourceTest.java
package ru.fix.dynamic.property.zk;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import ru.fix.dynamic.property.api.DynamicPropertySource;
import ru.fix.dynamic.property.jackson.JacksonDynamicPropertyMarshaller;
import ru.fix.dynamic.property.source.DefaultValue;
import ru.fix.dynamic.property.source.SourcedProperty;
import ru.fix.zookeeper.testing.ZKTestingServer;

import java.nio.charset.StandardCharsets;
import java.util.Map;
import java.util.Objects;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicReference;
import java.util.function.Supplier;

import static org.junit.jupiter.api.Assertions.*;


public class ZkDynamicPropertySourceTest {

    private static final String TEST_PROP_KEY = "test_prop_key";
    private static final String TEST_PROP_KEY_1 = "test_prop_key_1";
    private static final String PROPERTIES_LOCATION = "/zookeeper/p";

    private static final Integer TIMEOUT_SEC = 10;

    private ZKTestingServer zkTestingServer;

    @BeforeEach
    public void beforeEach() throws Exception {
        zkTestingServer = new ZKTestingServer();
        zkTestingServer.start();
    }

    @Test
    public void shouldIgnoreDefaultValueIfPropertyExists() throws Exception {
        DynamicPropertySource source = createPropertySource();

        CountDownLatch propertyChanged = new CountDownLatch(1);
        AtomicReference<String> newPropertySlot = new AtomicReference<>();

        DynamicPropertySource.Subscription sub = source.subscribeAndCallListener(
                TEST_PROP_KEY,
                String.class,
                DefaultValue.of("zzz"),
                newValue -> {
                    newPropertySlot.set(newValue);
                    propertyChanged.countDown();
                });

        setServerProperty(PROPERTIES_LOCATION + "/" + TEST_PROP_KEY, "some Value");

        propertyChanged.await();
        assertEquals("some Value", newPropertySlot.get());
    }

    @Test
    public void shouldFetchActualValueOfProperty() throws Exception {
        DynamicPropertySource source = createPropertySource();

        setServerProperty(PROPERTIES_LOCATION + "/" + TEST_PROP_KEY, "some Value");

        CountDownLatch firstChange = new CountDownLatch(1);
        CountDownLatch secondChange = new CountDownLatch(1);

        AtomicReference<String> valueSlot = new AtomicReference<>();

        DynamicPropertySource.Subscription sub = source.subscribeAndCallListener(
                TEST_PROP_KEY,
                String.class,
                DefaultValue.none(),
                (value) -> {
                    valueSlot.set(value);
                    if (1 == firstChange.getCount()) {
                        firstChange.countDown();
                    } else {
                        secondChange.countDown();
                    }
                });

        firstChange.await();
        assertEquals("some Value", valueSlot.get());

        changeServerProperty(PROPERTIES_LOCATION + "/" + TEST_PROP_KEY, "some Value 2");

        secondChange.await();
        assertEquals("some Value 2", valueSlot.get());
    }

    @Test
    public void shouldListenChangeOfProperty() throws Exception {
        DynamicPropertySource source = createPropertySource();

        CountDownLatch propertyAdd = new CountDownLatch(1);
        CountDownLatch propertyChanged = new CountDownLatch(1);

        AtomicReference<String> valueSlot = new AtomicReference<>();

        DynamicPropertySource.Subscription sub = source.subscribeAndCallListener(
                TEST_PROP_KEY_1,
                String.class,
                DefaultValue.none(),
                value -> {
                    valueSlot.set(value);
                    if (1 == propertyAdd.getCount()) {
                        propertyAdd.countDown();
                    }

                    if (1 == propertyChanged.getCount()) {
                        propertyChanged.countDown();
                    }
                });

        setServerProperty(PROPERTIES_LOCATION + "/" + TEST_PROP_KEY_1, "some Value");
        propertyAdd.await();

        changeServerProperty(PROPERTIES_LOCATION + "/" + TEST_PROP_KEY_1, "some Value 2");
        assertTrue(propertyChanged.await(TIMEOUT_SEC, TimeUnit.SECONDS));

        assertEquals("some Value 2", valueSlot.get());
    }

    @Test
    public void shouldListenRemovingOfProperty() throws Exception {
        DynamicPropertySource source = createPropertySource();

        CountDownLatch propertyAdd = new CountDownLatch(1);
        CountDownLatch propertyRemoved = new CountDownLatch(1);
        DynamicPropertySource.Subscription sub = source.subscribeAndCallListener(
                TEST_PROP_KEY_1,
                String.class,
                DefaultValue.of("default"),
                value -> {
                    if ("default".equals(value)) {
                        propertyRemoved.countDown();
                    } else {
                        propertyAdd.countDown();
                    }
                });

        setServerProperty(PROPERTIES_LOCATION + "/" + TEST_PROP_KEY_1, "some Value");
        propertyAdd.await();

        removeServerProperty(PROPERTIES_LOCATION + "/" + TEST_PROP_KEY_1);
        assertTrue(propertyRemoved.await(TIMEOUT_SEC, TimeUnit.SECONDS));
    }

    @Test
    public void shouldFetchAllProperties() throws Exception {
        ZkDynamicPropertySource source = createPropertySource();

        setServerProperty(PROPERTIES_LOCATION + "/propName1", "some Value 1");
        setServerProperty(PROPERTIES_LOCATION + "/propName2", "some Value 2");

        Map<String, Object> childProperties = source.readAllProperties();
        assertNotNull(childProperties);
        assertEquals(2, childProperties.size());
        assertEquals("some Value 1", childProperties.get("propName1"));
        assertEquals("some Value 2", childProperties.get("propName2"));
    }

    @Test
    public void shouldNotFetchPropertiesIfNotPresent() throws Exception {
        ZkDynamicPropertySource source = createPropertySource();

        Map<String, Object> allProperties = source.readAllProperties();
        assertNotNull(allProperties);
        assertEquals(allProperties.size(), 0);
    }

    @Test
    public void shouldGetActualValueFromHolder() throws Exception {
        ZkDynamicPropertySource source = createPropertySource();

        String propertyNewValue = "some Value 2";

        CountDownLatch propertyAdded = new CountDownLatch(1);
        CountDownLatch propertyRemoved = new CountDownLatch(1);

        DynamicPropertySource.Subscription sub = source.subscribeAndCallListener(
                "propName1",
                String.class,
                DefaultValue.of("default"),
                value -> {
                    if (value.equals("default")) {
                        propertyRemoved.countDown();
                    } else {
                        propertyAdded.countDown();
                    }
                });

        SourcedProperty<String> propertyHolder = new SourcedProperty<>(
                source,
                "propName1",
                String.class,
                DefaultValue.none());

        retryAssert(null, propertyHolder::get);

        setServerProperty(PROPERTIES_LOCATION + "/propName1", propertyNewValue);
        propertyAdded.await();
        retryAssert(propertyNewValue, propertyHolder::get);

        removeServerProperty(PROPERTIES_LOCATION + "/propName1");
        propertyRemoved.await();
        retryAssert(null, propertyHolder::get);
    }

    public <T> void retryAssert(T expectedValue, Supplier<T> actualVale) throws InterruptedException {
        long assertEndTime = System.currentTimeMillis() + TIMEOUT_SEC;

        while (System.currentTimeMillis() <= assertEndTime) {
            if (Objects.equals(expectedValue, actualVale.get())) {
                break;
            }
            TimeUnit.MICROSECONDS.sleep(100);
        }

        assertEquals(expectedValue, actualVale.get());
    }

    @Test
    public void shouldGetDefaultValueFromHolder() throws Exception {
        ZkDynamicPropertySource source = createPropertySource();

        SourcedProperty<String> holder = new SourcedProperty<String>(
                source,
                "unknown.property",
                String.class,
                DefaultValue.of("default Value")
        );
        assertEquals("default Value", holder.get());
    }

    private ZkDynamicPropertySource createPropertySource() throws Exception {
        return new ZkDynamicPropertySource(
                zkTestingServer.getClient(),
                PROPERTIES_LOCATION,
                new JacksonDynamicPropertyMarshaller()
        );
    }

    private void setServerProperty(String propertyKey, String value) throws Exception {
        zkTestingServer.getClient().create().creatingParentsIfNeeded().forPath(
                propertyKey,
                value.getBytes(StandardCharsets.UTF_8)
        );
    }

    private void changeServerProperty(String propertyKey, String value) throws Exception {
        zkTestingServer.getClient().setData().forPath(propertyKey, value.getBytes(StandardCharsets.UTF_8));
    }

    private void removeServerProperty(String propertyKey) throws Exception {
        zkTestingServer.getClient().delete().deletingChildrenIfNeeded().forPath(propertyKey);
    }
}

File: dynamic-property-zk/src/test/kotlin/ru/fix/dynamic/property/zk/ZkDynamicPropertySourceTest.kt
package ru.fix.dynamic.property.zk

import org.awaitility.Awaitility.await
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import ru.fix.dynamic.property.api.source.OptionalDefaultValue
import ru.fix.dynamic.property.api.source.SourcedProperty
import ru.fix.dynamic.property.jackson.MarshallerBuilder
import ru.fix.zookeeper.testing.ZKTestingServer
import java.nio.charset.StandardCharsets
import java.time.Duration
import java.time.temporal.ChronoUnit
import java.util.concurrent.LinkedBlockingDeque
import java.util.concurrent.TimeUnit


class ZkDynamicPropertySourceTest {
    companion object {
        private const val TEST_PROP_KEY = "test_prop_key"
        private const val TEST_PROP_KEY_1 = "test_prop_key_1"
        private const val PROPERTIES_LOCATION = "/zookeeper/p"
    }

    private lateinit var zkTestingServer: ZKTestingServer
    private lateinit var source: ZkDynamicPropertySource

    @BeforeEach
    fun beforeEach() {
        zkTestingServer = ZKTestingServer().start()
        source = ZkDynamicPropertySource(
            zkTestingServer.client,
            PROPERTIES_LOCATION,
            MarshallerBuilder().build(),
            Duration.of(1, ChronoUnit.MINUTES)
        )
    }

    @AfterEach
    fun afterEach() {
        source.close()
        zkTestingServer.close()
    }

    @Test
    fun `property exist in store, ignore default value`() {
        val slot = LinkedBlockingDeque<String>()

        setServerProperty("$PROPERTIES_LOCATION/$TEST_PROP_KEY", "some Value")

        val subscription = source.createSubscription(
            TEST_PROP_KEY,
            String::class.java,
            OptionalDefaultValue.of("zzz")
        ).setAndCallListener { slot.add(it) }

        assertEquals("some Value", slot.removeFirst())
        assertTrue(slot.isEmpty())
    }

    @Test
    fun `property exist in store, ignore default value, react on property update`() {
        setServerProperty("$PROPERTIES_LOCATION/$TEST_PROP_KEY", "some Value")

        val valueSlot = LinkedBlockingDeque<String>()

        val subscription = source.createSubscription(
            TEST_PROP_KEY,
            String::class.java,
            OptionalDefaultValue.of("zzz")
        ).setAndCallListener { value -> valueSlot.add(value) }

        assertEquals("some Value", valueSlot.removeFirst())

        changeServerProperty("$PROPERTIES_LOCATION/$TEST_PROP_KEY", "some Value 2")

        assertEquals("some Value 2", valueSlot.takeFirst())
    }


    @Test
    fun `start with default value and then listen for property creation and change`() {
        val valueSlot = LinkedBlockingDeque<String>()

        val subscription = source.createSubscription(
            TEST_PROP_KEY_1,
            String::class.java,
            OptionalDefaultValue.of("zzz")
        ).setAndCallListener { value ->
            valueSlot.add(value)
        }

        assertEquals("zzz", valueSlot.takeFirst())
        assertTrue(valueSlot.isEmpty())

        setServerProperty("$PROPERTIES_LOCATION/$TEST_PROP_KEY_1", "some Value")
        assertEquals("some Value", valueSlot.takeFirst())
        assertTrue(valueSlot.isEmpty())

        changeServerProperty("$PROPERTIES_LOCATION/$TEST_PROP_KEY_1", "some Value 2")
        assertEquals("some Value 2", valueSlot.takeFirst())
        assertTrue(valueSlot.isEmpty())
    }

    @Test
    fun `property removed from source, use default value`() {
        val slot = LinkedBlockingDeque<String>()

        val sub = source.createSubscription(
            TEST_PROP_KEY_1,
            String::class.java,
            OptionalDefaultValue.of("default")
        ).setAndCallListener { value ->
            slot.add(value)
        }

        assertEquals("default", slot.removeFirst())
        assertTrue(slot.isEmpty())

        setServerProperty("$PROPERTIES_LOCATION/$TEST_PROP_KEY_1", "some Value")
        assertEquals("some Value", slot.takeFirst())
        assertTrue(slot.isEmpty())

        removeServerProperty("$PROPERTIES_LOCATION/$TEST_PROP_KEY_1")
        assertEquals("default", slot.takeFirst())
        assertTrue(slot.isEmpty())
    }

    @Test
    fun `read all properties`() {
        setServerProperty("$PROPERTIES_LOCATION/propName1", "some Value 1")
        setServerProperty("$PROPERTIES_LOCATION/propName2", "some Value 2")

        val childProperties = source.readAllProperties()
        assertNotNull(childProperties)
        assertEquals(2, childProperties.size)
        assertEquals("some Value 1", childProperties["propName1"])
        assertEquals("some Value 2", childProperties["propName2"])
    }

    @Test
    fun shouldNotFetchPropertiesIfNotPresent() {
        val allProperties = source.readAllProperties()
        assertNotNull(allProperties)
        assertEquals(allProperties.size, 0)
    }

    @Test
    fun `default value of one holder does not affect default value of another holder`() {

        val propertyHolder = SourcedProperty(
            source,
            "propName1",
            String::class.java,
            OptionalDefaultValue.of("defaultHolderValue")
        )

        val propertyHolder2 = SourcedProperty(
            source,
            "propName1",
            String::class.java,
            OptionalDefaultValue.of("defaultHolderValue2")
        )

        assertEquals("defaultHolderValue", propertyHolder.get())

        setServerProperty("$PROPERTIES_LOCATION/propName1", "some Value 2")

        await().atMost(10, TimeUnit.SECONDS).until {
            propertyHolder.get() == "some Value 2"
        }

        removeServerProperty("$PROPERTIES_LOCATION/propName1")

        await().atMost(10, TimeUnit.SECONDS).until {
            propertyHolder.get() == "defaultHolderValue"
        }
    }


    @Test
    fun shouldGetDefaultValueFromHolder() {
        val holder = SourcedProperty(
            source,
            "unknown.property",
            String::class.java,
            OptionalDefaultValue.of("default Value")
        )
        assertEquals("default Value", holder.get())
    }

    @Test
    fun `should return actual value instead of default when a large number of properties loaded`() {
        val generatedProperties = generateProperties(200)

        generatedProperties.forEach {
            setServerProperty("${PROPERTIES_LOCATION}/${it.key}", it.value)
        }

        val source = ZkDynamicPropertySource(
            zkTestingServer.createClient(),
            PROPERTIES_LOCATION,
            MarshallerBuilder().build(),
            Duration.of(1, ChronoUnit.MINUTES)
        )

        val subscriptions = generatedProperties.map {
            source.createSubscription(
                it.key,
                String::class.java,
                OptionalDefaultValue.of("default")
            ).setAndCallListener { value ->
                assertNotEquals("default", value)
            }
        }

        source.close()
    }

    private fun generateProperties(count: Int): Map<String, String> {
        return (1..count).map { i -> Pair("prop-$i", "value-$i") }.toMap()
    }

    private fun setServerProperty(propertyKey: String, value: String) {
        val data = value.toByteArray(StandardCharsets.UTF_8)
        zkTestingServer.client
            .create()
            .creatingParentsIfNeeded()
            .forPath(propertyKey, data)

        await().atMost(10, TimeUnit.SECONDS).until {
            zkTestingServer.client
                .data
                .forPath(propertyKey)!!.contentEquals(data)
        }
    }

    private fun changeServerProperty(propertyKey: String, value: String) {
        zkTestingServer.client.setData().forPath(propertyKey, value.toByteArray(StandardCharsets.UTF_8))
    }


    private fun removeServerProperty(propertyKey: String) {
        zkTestingServer.client.delete().deletingChildrenIfNeeded().forPath(propertyKey)
    }
}
